### PR TITLE
remove user-facing undefined in preference to Proxy

### LINKED ISSUE #144: Consider using proxy types instead of undefined in some places
As I've been working more with `Tupleable`, etc, I've realized some differences (and potential pitfalls) of using `undefined :: a` instead of `Proxy @a` (or `Proxy :: Proxy a` for GHC < 8):

Unlike `Proxy`, `undefined` masquerades as a *value* of type `a`. If your API doesn't touch this value, then, of course,  you're fine. However, if you have any available constraints on `a`, then all of those type-class methods will work. For example, `Tupleable a` has `NFData a` and `Binary a`, which means the implementation of `Tupleable` can use methods of these classes, possibly inadvertently.

The benefit of a proxy type here is that you simply can't implement any of this incorrectly, because you never actually get a value of type `a`.

```haskell
class Tupleable a where
  toAttributes :: proxy a -> [Attribute]
  ...

-- call-sight
toAttributes (Proxy @MyRecord)
```

It helps to know that you can make a type variable available in the body of a function using `-XScopedTypeVariables` and `forall`.

### PR DESCRIPTION
**Pull request by agentm**
Is it OK to continue to use `undefined` under the hood?

resolves #144

### CODE REVIEW

**Comment by cumber** [comment]
Yeah, nobody cares how the sausage is made. ;)

What did you end up deciding to do about relations as atoms? Is that still usable without them being `Atomable`?

**Comment by agentm** [comment]
I agree that the use case for `Relation`s within data types that users will care about is slim, so the default `Atomable` instance for `Relation`s has been removed. If a user needs such an instance, he can create his own orphan instance with a pre-cooked `AtomType`.

**Comment by 3noch** [comment]
I'm not sure all the implications but you can actually make the proxy type polymorphic because it doesn't actually matter what type carries the type.

**Comment by 3noch** [comment]
The advantage is that by using `forall a proxy. proxy a` you are swearing to the world that you can't possibly inspect either the `a` *or* the `proxy`.

**Comment by 3noch** [comment]
Note that I prefer `forall a proxy` to `forall proxy a` because type applications can be partially applied and you usually prefer to specify the type being carried rather than than the type carrying it.

E.g.

```haskell
toAtomType :: forall a proxy. proxy a -> AtomType`
...
toAtomType @Int

-- vs.

toAtomType' :: forall proxy a. proxy a -> AtomType`
...
toAtomType' _ @Int
```

**Comment by 3noch** [comment]
I'm sorry, I forgot that at the edges of your API you *do* need to use an explicit proxy type because you need to actually pattern match on it (and the caller must create it). However, inside your API you often just pass the type around to other parts of the API and in these cases you can be polymorphic. Though at that point I'm not sure what it buys you.

**Comment by 3noch** [comment]
Just to be clear: Good work. I like these changes.

File: docs/tupleable.markdown [hunk 1]
```diff
@@ -27,9 +27,9 @@ Now, ```Person``` values can be marshaled between the database and the client. T
 
 The ```Tupleable``` module includes the following functions:
 
- * ```toInsertExpr :: (Tupleable a, Traversable t) => t a -> RelVarName -> Either RelationalError DatabaseContextExpr```
+ * ```toInsertExpr :: forall a. forall t. (Tupleable a, Traversable t) => t a -> RelVarName -> Either RelationalError DatabaseContextExpr```
```
Comment (3noch) [inline docs/tupleable.markdown]: Can this be `forall a t.`?

**Pull request merged**